src/main/java/ui/components/pickers/MilestonePickerState.java

```diff
@@ -0,0 +1,145 @@
+package ui.components.pickers;
+
+import util.Utility;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+public class MilestonePickerState {
+    private List<PickerMilestone> currentMilestonesList;
+
+    public MilestonePickerState(List<PickerMilestone> milestones) {
+        currentMilestonesList = cloneList(milestones);
+    }
+
+    public MilestonePickerState(List<PickerMilestone> milestones, String userInput) {
+        this(milestones);
+        processInput(userInput);
+    }
+
+    private List<PickerMilestone> cloneList(List<PickerMilestone> sourceList) {
+        List<PickerMilestone> destList = new ArrayList<>();
+        sourceList.stream()
+                .forEach(milestone -> destList.add(new PickerMilestone(milestone)));
+        return destList;
+    }
+
+    private void processInput(String userInput) {
+        if (userInput.isEmpty()) return;
+
+        String[] userInputWords = userInput.split(" ");
+        for (int i = 0; i < userInputWords.length; i++) {
+            String currentWord = userInputWords[i];
+            if (isConfirmedKeyword(userInput, userInputWords, i)) {
+                toggleFirstMatchingMilestone(currentWord);
+            } else {
+                filterMilestones(currentWord);
+            }
+        }
+    }
+
+    /**
+     * Checks if the ith word of the userInput is confirmed i.e. has a space after it
+     * @param userInput
+     * @param userInputWords
+     * @param i
+     * @return
+     */
+    private boolean isConfirmedKeyword(String userInput, String[] userInputWords, int i) {
+        return i < userInputWords.length - 1 || userInput.endsWith(" ");
+    }
+
+    /**
+     * Toggles the milestone that has milestoneName as its exact name
+     *
+     * This method is case-sensitive
+     * @param milestoneName
+     */
+    public final void toggleExactMatchMilestone(String milestoneName) {
+        Optional<PickerMilestone> onlyMatchingMilestone = getExactMatchMilestone(currentMilestonesList, milestoneName);
+        toggleMilestone(onlyMatchingMilestone);
```

Comment:
Rather than call this with an optional, we should check here if it's present and call/not call it here.